operator/src/main/java/org/keycloak/operator/v2alpha1/WatchedSecretsStore.java

```diff
@@ -0,0 +1,237 @@
+    }
+
+    private String getSecretVersion(Secret secret) {
+        String serializedData = Serialization.asYaml(secret.getData());
+        try {
+            // using hashes as it's more robust than resource versions that can change e.g. just when adding a label
+            byte[] bytes = MessageDigest.getInstance("MD5").digest(serializedData.getBytes(StandardCharsets.UTF_8));
+            return new BigInteger(1, bytes).toString(16);
+        }
+        catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private Set<Secret> fetchCurrentSecrets(Set<String> secretsNames) {
+        return secretsNames.stream()
+                .map(n -> {
+                    Secret secret = client.secrets().inNamespace(getNamespace()).withName(n).get();
+                    if (secret == null) {
+                        throw new IllegalStateException("Secret " + n + " not found");
+                    }
+                    return secret;
+                })
+                .collect(Collectors.toSet());
+    }
+
+    @Override
+    public String getName() {
+        return cr.getMetadata().getName() + STORE_SUFFIX;
+    }
+
+    public static EventSource getStoreEventSource(KubernetesClient client, String namespace) {
+        SharedIndexInformer<Secret> informer =
+                client.secrets()
+                        .inNamespace(namespace)
+                        .withLabel(Constants.TYPE_LABEL, TYPE)
+                        .runnableInformer(0);
+
+        return new InformerEventSource<>(informer, Mappers.fromOwnerReference()) {
+            @Override
+            public String name() {
+                return "watchedResourcesStoreEventSource";
+            }
+        };
+    }
+
+    public static EventSource getWatchedSecretsEventSource(KubernetesClient client, String namespace) {
+        SharedIndexInformer<Secret> informer =
+                client.secrets()
+                        .inNamespace(namespace)
+                        .withLabel(Constants.TYPE_LABEL, TYPE_WATCHED)
+                        .runnableInformer(0);
+
+        return new InformerEventSource<>(informer, secret -> {
+            // get all stores
+            List<Secret> stores = client.secrets().inNamespace(namespace).withLabel(Constants.TYPE_LABEL, TYPE).list().getItems();
+
+            // find all CR names that are watching this Secret
+            var ret = stores.stream()
+                    // check if any of the stores tracks this secret
+                    .filter(store -> store.getData().containsKey(secret.getMetadata().getName()))
+                    .map(store -> {
+                        String crName = store.getMetadata().getName().split(STORE_SUFFIX)[0];
+                        return new ResourceID(crName, namespace);
+                    })
+                    .collect(Collectors.toSet());
+
+            if (ret.isEmpty()) {
```

Comment:
I understand why you did it here, and the implementation is straightforward, but I need to ask you to separate this functionality from the informer computation, something on those lines:

```java
public static void cleanObsoleteLabels() {
  client
    .secrets()
    .inNamespace(namespace)
    .withLabel(Constants.TYPE_LABEL, TYPE)
    .list()
    .getItems()
    .filter(store -> client.resources(Keycloak.class).inNamespace(namespace).withName(store.getMetadata().getName().split(STORE_SUFFIX)[0]).get() == null)
    .forEach(s -> {... remove the label ...})
}
```

Reply:
Will do, but not to unnecessarily fetch all the Secrets all the time, can't separate it completely.

Reply:
not really excited by the current implementation, but we can iterate over it in the future.